File: chapter1/milne.py
import math
import numpy as np
import matplotlib.pyplot as plt

# Constantes globales 
Nz = 60
Nmu = 100
kmax = 15
tau1 = 1.0
dtau = tau1 / (Nz - 1)
mus = 1.0 / math.sqrt(3.0) # Une inclinaison de 60°
Is = 1.0 # Intensité de la lumière du soleil

# Initialisation des tableaux
J00 = [0.0] * Nz
J0 = [0.0] * Nz
S = [0.0] * Nz

def sqr(x):
    """Fonction utilitaire équivalente à la macro #define sqr(x)"""
    return x * x

def expint_E1(t=1.0):
    """Précision dans la fonction d'intégrale exponentielle E1"""
    t1 = abs(t) # permet les appels avec un argument négatif
    Kexpint = int(9 + (t1 - 1) * 4) 
    gaNtaua = 0.577215664901533 # intégration spéciale pour log(t)
    
    if t1 < 1e-5:
        return 0.0 # E1(t), t petit est toujours appelé avec t*E1(t)
    
    if t1 > 4.0:
        tx = 1.0 / t1
        return math.exp(-t1) * tx * (1 + (-1 + (2 + (-6 + (24 + (-120 + 720 * tx) * tx) * tx) * tx) * tx) * tx)
    
    ak = t1
    soNtaue = -gaNtaua - math.log(t1) + ak
    for k in range(2, Kexpint):
        ak *= -t1 * (k - 1) / sqr(k)
        soNtaue += ak
        
    return soNtaue

def getI():
    pi = 4.0 * math.atan(1.0)
    dmu = 2.0 / Nmu
    b = 1 # 1 pour Chandrasekhar, 0 pour Dirac
    lam = 12.0
    w = math.sqrt(pi) / 2.0 / lam * (math.erf(mus * lam) + math.erf((1.0 - mus) * lam))
    
    print(" tau,mu --> I(tau,mu) ")
    
    mu = -1.0
    while mu < 1.0:
        for j in range(Nz):
            # En Python, int(True) vaut 1 et int(False) vaut 0, ce qui remplace la syntaxe (mu<0) du C++
            Iss = int(mu < 0) * Is * math.exp(-sqr((mu + mus) * lam)) * math.exp(-j * dtau / mus) / w
            I = (1 - b) * Iss
            
            for i in range(Nz):
                condition = int(i < j) * int(mu < 0) + int(i > j) * int(mu > 0)
                if abs(mu) > 0: # Évite la division par zéro
                    I += condition * dtau * math.exp(-abs((i - j) * dtau / abs(mu))) \
                         * (J0[i] + b * Is * math.exp(-i * dtau / mus) / 2.0) / abs(mu)
                         
            print(f"{j * dtau:.5f}\t{mu:.5f}\t{I + b * Iss:.5f}")
        print()
        mu += dmu
        
    return 0
def getI2():
    pi = 4.0 * math.atan(1.0)
    dmu = 2.0 / Nmu
    b = 1  # 1 pour Chandrasekhar, 0 pour Dirac
    lam = 12.0
    w = math.sqrt(pi) / 2.0 / lam * (math.erf(mus * lam) + math.erf((1.0 - mus) * lam))
    
    # Listes pour stocker les données du graphique
    mu_vals = []
    tau_vals = [j * dtau for j in range(Nz)]
    I_surface = []
    
    print("Calcul de la surface I(tau, mu) en cours...")
    
    mu = -1.0
    while mu <= 1.0:
        mu_vals.append(mu)
        I_row = []
        
        for j in range(Nz):
            Iss = int(mu < 0) * Is * math.exp(-sqr((mu + mus) * lam)) * math.exp(-j * dtau / mus) / w
            I = (1 - b) * Iss
            
            for i in range(Nz):
                condition = int(i < j) * int(mu < 0) + int(i > j) * int(mu > 0)
                if abs(mu) > 1e-9:  # Sécurité pour éviter la division par zéro
                    I += condition * dtau * math.exp(-abs((i - j) * dtau / abs(mu))) \
                         * (J0[i] + b * Is * math.exp(-i * dtau / mus) / 2.0) / abs(mu)
                         
            I_row.append(I + b * Iss)
            
        I_surface.append(I_row)
        mu += dmu

    # Conversion en tableaux NumPy pour la reconstruction de la surface
    Mu, Tau = np.meshgrid(mu_vals, tau_vals, indexing='ij')
    I_surface = np.array(I_surface)
    
    # --- Création du graphique 3D ---
    fig = plt.figure(figsize=(12, 8))
    ax = fig.add_subplot(111, projection='3d')
    
    # Tracé de la surface
    surf = ax.plot_surface(Tau, Mu, I_surface, cmap='viridis', edgecolor='none', alpha=0.9)
    
    # Personnalisation des axes
    ax.set_xlabel(r'$\tau$ (Profondeur optique)', fontsize=12)
    ax.set_ylabel(r'$\mu$ ($\cos(\theta)$)', fontsize=12)
    ax.set_zlabel(r'$I(\tau, \mu)$', fontsize=12)
    ax.set_title("Surface de l'intensité spécifique $I(\tau, \mu)$ (Problème de Milne)", fontsize=14, pad=20)
    
    # Ajout d'une barre de couleur
    fig.colorbar(surf, ax=ax, shrink=0.5, aspect=10, label="Intensité $I$")
    
    # Orientation initiale de la vue 3D (ajustable à la souris)
    ax.view_init(elev=30, azim=-60)
    
    plt.tight_layout()
    plt.show()
    return 0

def schwarzschild(tau):
    J00_val = Is / 2.0
    B = J00_val / tau1 / 2.0
    A = J00_val - B
    exp2tau = math.exp(-2.0 * tau)
    return 2.0 * (J00_val * exp2tau + A * (1.0 - exp2tau) - B * (2.0 * tau - 1.0 + exp2tau))

def main():
    global J0, S
    
    # Initialisation
    for i in range(Nz):
        J0[i] = math.exp(-i * dtau / mus) * Is / 2.0
        
    print("k     J0[1]")
    
    # Itérations sur la source (point fixe)
    for k in range(kmax):
        for i in range(Nz):
            S[i] = J0[i]
            
        for i in range(Nz):
            J0[i] = math.exp(-i * dtau / mus) * Is / 2.0
            for j in range(1, Nz):
                J0[i] += dtau / 4.0 * expint_E1((j - i - 0.5) * dtau) * (S[j] + S[j - 1])
                
        print(f"{k} {J0[1]}")
        
    print("\ntau      J0[tau]      Schwarzschild")
    for i in range(Nz):
        print(f"{i * dtau:.5f} {J0[i]:.5f} {schwarzschild(i * dtau):.5f}")
        
getI2() # Uncomment to display  I(tau,mu)

if __name__ == "__main__":
    main()

File: chapter1/Milne.cpp
// Solves Milne's Problem
// compile with g++ -O3 -std=c++17  -o Milne MilneProblem.cpp

#include <iostream>
#include <fstream>
#include <cmath>
#include <string>
#include <time.h>
using namespace std;
#define sqr(x) ((x)*(x))

const int Nz=60, Nmu=100; // nb points in z and tau in (0,1)
const int kmax=15;  // nb fixed point iterations
const double tau1=2;
const double dtau=tau1/(Nz-1);
const double mus=1./sqrt(3.); // an inclination is 60°
const double Is= 1; // Sunlight intensity
double J0[Nz], S[Nz];

double expint_E1(const double t=1){
     double t1=fabs(t); // allow calls with negative argument
    const int Kexpint = 9+(t1-1)*4; // precision in exponential integral function E1
    const double  gaNtaua =0.577215664901533; // special integration for log(t)
    if(t1<1e-5) return 0; // E1(t), t small is always called with t*E1(t)
    if(t1>4) {
        double tx=1./t1;
        return exp(-t1)*tx * (1 +(-1+(2+(-6+(24+(-120+720*tx)*tx)*tx)*tx)*tx)*tx );
    }
    double ak=t1, soNtaue=-gaNtaua - log(t1)+ak;
    for(int k=2;k<Kexpint;k++){
        ak *= -t1*(k-1)/sqr(k);
        soNtaue += ak;
    }
    return soNtaue;
}

int getI(){
    const double pi=4*atan(1.), dmu=2./Nmu;
    const int b=1 ; // 1 for Chandrasekhar, 0 for Dirac
    const double lam=12,  w = sqrt(pi)/2/lam * ( erf(mus*lam) + erf((1-mus)*lam) ) ;
    cout<<" tau,mu --> I(tau,mu) \n";
    for(double mu=-1; mu<1;mu+=dmu){
        for(int j=0;j<Nz;j++){
            double Iss= (mu<0)*Is*exp(-sqr((mu+mus)*lam))*exp(-j*dtau/mus)/w, I=(1-b)*Iss;
            for(int i=0;i<Nz;i++)
                I +=( (i<j)*(mu<0) + (i>j)*(mu>0) )* dtau*exp(-fabs((i-j)*dtau/fabs(mu)))
                        * ( J0[i] + b*Is*exp(-i*dtau/mus)/2 )/fabs(mu);
            cout<<j*dtau<<"\t"<<mu<<"\t"<< I+ b*Iss<< endl;
        }
        cout<<endl;
    }
    return 0;
}

double schwarzschild(const double tau){
    const double J00=Is/2;
    return J00*(1-(1+tau)/(1+tau1)) +J00*(1-tau/(1+tau1));
}
int main(int argc, const char * argv[]) {
    
    for(int i=0;i<Nz;i++) // initialization
        J0[i] =  exp(-i*dtau/mus)*Is/2;
    
    cout<<"k     J0[1]"<<endl;
    for(int k=0;k<kmax;k++){ // iterations on the source
        for(int i=0;i<Nz;i++)
            S[i] = J0[i] ;
        for(int i=0;i<Nz;i++) {
            J0[i] = exp(-i*dtau/mus)*Is/2;
            for(int j=1;j<Nz;j++)
                J0[i] += dtau/4 * expint_E1((j-i-0.5)*dtau)*(S[j]+S[j-1]);
        }
        cout<<k<<" "<<J0[1]<<endl;
    }
    cout<<endl<<"tau      J0[tau]   Schwarzschild[tau]"<<endl;
    for(int i=0;i<Nz;i++) cout<< i*dtau<<" "<< J0[i]<<" "<< schwarzschild(i*dtau) << endl;
 //   getI(); //uncomment to display I(tau,mu)
    return 0;
}
